### ISSUE TITLE
SnapError imported and used inside of a /pkg/cfgfile

### ISSUE THREAD

**Issue by geauxvirtual** [intelsdi-x/snap#903]
SnapError imported and used inside of /pkg/cfgfile. Code existing inside of /pkg directory should have no dependencies on Snap.

**Comment by mjbrender** [comment]
@geauxvirtual fair to label as bug to fix this?

**Comment by IRCody** [comment]
What is the reasoning behind the pkg directory not using anything from snap?

**Comment by geauxvirtual** [comment]
These packages provide functionality that could be consumed by other projects. If I wanted to import pkg/cfgfile into another package, I shouldn't also have to import core/serror just to convert SnapErrors back to regular errors.

We have maintained from the beginning of Snap that anything in the pkg directory can not import anything from snap.

**Comment by IRCody** [comment]
>These packages provide functionality that could be consumed by other projects. If I wanted to import pkg/cfgfile into another package, I shouldn't also have to import core/serror just to convert SnapErrors back to regular errors.

If they are meant to be completely separate wouldn't it be more intuitive/conventional to have them live in separate repos and import them that way?

**Comment by geauxvirtual** [comment]
We've had no reason to separate them into a separate repo, and many other projects follow the same pattern.

**Comment by lynxbat** [comment]
The rule for `\pkg` was potential packages that would land in their own repo later. @geauxvirtual is right in that either it should not use `serror` or `serror` should be moved to `\pkg`.

**Comment by tjmcs** [comment]
I'm happy to refactor to get rid of the dependency in my currently open PR (PR #904) or open a new one, @geauxvirtual and @lynxbat; your choice...

Personally, I don't think moving the `SnapError` code into the `pkg` package makes sense if that package is intended to be something that can be used independently of Snap, but I could be convinced otherwise if anyone else feels strongly about it.

The fact that it was included in that method was simply because we, @IRCody and I, were unfamiliar with that rule (that `pkg` should not have any Snap dependencies) and there were others from the team (who shall remain nameless) that recommended that we modify those routines to return a SnapError instead of a slice of errors (to make display of the errors found in a configuration file easier to manage). It's easy enough to modify, and it'll touch the same files I've been modifying in that PR.

**Comment by lynxbat** [comment]
Yes. I think it is best to stick with our rules on /pkg. Another option is to just move /cfgfile out of /pkg since it is to this point not imported elsewhere (unless I am wrong).

That said: the serror could just be wrapped around the error in the places inside Snap that call cfgfile.

Either is fine for me. I actually prefer @jcooklin chooses the path.

**Comment by jcooklin** [comment]
@tjmcs: Let's move it.

**Comment by tjmcs** [comment]
Where should we move it to?

**Comment by jcooklin** [comment]
core/cfgfile/...

**Comment by tiffanyfj** [comment]
^That's where I was thinking too

**Comment by geauxvirtual** [comment]
This functionality has zero to do with snap. A binary, snapd, that is written to run the snap control and scheduler uses cfgfile to read configuration files. snapctl, another binary, for interacting with snap REST API, could also use this functionality to read configuration files, if we so choose. But this is code that has nothing to do with snap, which is a controller of plugins and scheduler of tasks and a REST API.

**Comment by tjmcs** [comment]
Actually, I think I agree with @geauxvirtual; the functionality of this package is fairly generic (as is evidenced by the tests that I've been refactoring which have nothing to do with Snap and everything to do with reading configuration files containing either JSON or YAML into a fairly generic `interface{}`). So if no one objects, I think I should refactor the existing package (in place) to remove the `SnapError` dependency and return a slice of `error` instances instead...thoughts, @jcooklin and @lynxbat?

**Comment by lynxbat** [comment]
Let's go with @jcooklin's call. At this point and time this package only has to do with Snap. Moving it is the minimum effort and risk. We can move the isolation out till we need it on a project.
